tools/clang/lib/Sema/SemaHLSL.cpp

```diff
@@ -8882,6 +8883,84 @@ static NameLookup GetSingleFunctionDeclByName(clang::Sema *self, StringRef Name,
   return NameLookup{ pFoundDecl, nullptr };
 }
 
+namespace {
+/// A visitor class for validating HLSL attributes and annotations
+class HLSLAttributeChecker : public RecursiveASTVisitor<HLSLAttributeChecker> {
+public:
+  explicit HLSLAttributeChecker(Sema &S);
+
+  bool TraverseFunctionDecl(FunctionDecl* func);
+
+  bool VisitFunctionDecl(FunctionDecl *func);
+  bool VisitParmVarDecl(ParmVarDecl *parmVar);
+
+private:
+  typedef RecursiveASTVisitor<HLSLAttributeChecker> BaseType;
+
+  Sema &theSema;
+  clang::DiagnosticsEngine &theDiag;
+  const hlsl::ShaderModel *shaderModel;
+
+  bool isEntryFunction;
+};
+
+HLSLAttributeChecker::HLSLAttributeChecker(Sema &S)
+    : theSema(S), theDiag(S.getDiagnostics()), shaderModel(nullptr),
+      isEntryFunction(false) {
+  shaderModel =
+      hlsl::ShaderModel::GetByName(theSema.getLangOpts().HLSLProfile.c_str());
+}
+
+bool HLSLAttributeChecker::TraverseFunctionDecl(FunctionDecl* func) {
+  if (!VisitFunctionDecl(func))
+    return false;
+
+  for (auto* param : func->params())
+    if (!TraverseParmVarDecl(param))
+      return false;
+
+  // Avoid traversing the statements
+
+  return true;
+}
+
+bool HLSLAttributeChecker::VisitFunctionDecl(FunctionDecl *func) {
+  isEntryFunction = theSema.getLangOpts().HLSLEntryFunction == func->getName();
+
+  if (isEntryFunction && shaderModel->IsGS()) {
```

Comment:
This is definitely complicated by the library case, and for patch constant function case if we add validation for that.

Reply:
Seems the maxvertexcount check can easily be made in DiagnoseTranslationUnit (given the HLSLProfile addition).

Reply:
Yes. Moved the check here.